### PR TITLE
Save and load prompt style automatically

### LINKED ISSUE #825: generate/lora is tied to Alpaca instruction style
The script is tied to Alpaca prompt style. I can understand that basically all the prepare scripts enforces Alpaca style, however  `generate/base` is agnostic (for good!)

[    prompt = generate_prompt(sample)](https://github.com/Lightning-AI/lit-gpt/blob/74b8df0c3f07fc31d9d1a49e870a1f7955329ad8/generate/lora.py#L101)

### LINKED ISSUE #954: Data Refactor Proposal
This issue proposes a refactor for how data is preprocessed and consumed in Lit-GPT.

## Current Issues and Limitations

**Error prone:** We have scripts in `scripts/prepare_*` to preprocess the data before running the scripts. This is cumbersome and error prone because you have to specify the tokenizer that you are going to use to finetune. If you then want to finetune a different model, and forget to rerun the prepare script, you will use wrongly tokenized data.

**Data overlap:** The way the finetuning scripts load the data is with a random index into the memory mapped preprocessed file (sampling with replacement). Training over N epochs isn't really possible, but for finetuning we often want to control this very precisely. In addition, distributed sampling is also an issue, and the current workaround is to set a different seed per rank but users are likely oblivious to this immense technical detail.

**Inflexible:** There is no standard interface to read a dataset in our scripts, making it harder to adopt new dataset types (e.g. DPO). Furthermore, the prompt template is hardcoded into the data preparation. 

**No code experience**: Lit-GPT is moving toward a CLI-focused experience where everything needs to be configurable without changing the code. The data part is the largest piece standing in the way of this at the moment.

## Proposed Changes
1. Switch all dataloading to PyTorch Dataset + DataLoader
2. Bundle all data related logic in a datamodule
3. Training scripts will instantiate a datamodule and call `prepare_data()`, `setup()`, and `train/va_dataloader()` to get the loaders (see also extension section below)
4. Predefined datamodules are available under `lit_gpt.datasets`
5. All datamodule related arguments are exposed in the training script CLI via `--data.xyz` arguments.
6. Our predefined datamodules are registered under a shortcut name that can be referenced via the CLI (e.g. `--data.module="Alpaca"`, making it super quick to change between datasets.
7. We can provide a generic datamodule for CSV, JSON etc. that can read user defined datasets in standardized format. E.g. `--data.module=csv --data.source="path/to/csv/or/folder"`
8. The prompt template is configureable

## Usage Examples
```py

# Use defaults provided by data module
python finetune/full.py --data.module=LIMA 

# Provide optional arguments
python finetune/full.py --data.module=LIMA --data.test_split_fraction=0.1
python finetune/full.py --data.module=LIMA --data.max_seq_length=256
```

## DataModule and Dataset

The DataModule could simply follow the `LightningDataModule` design, or even subclass it:

```py
class Alpaca:
    def __init__(
        self,
        max_seq_length: int = -1,
        mask_prompt: bool = True,
        test_split_fraction: float = 0.03865,
        # ...
    ) -> None:
        super().__init__()
        # ...

    def prepare_data(self) -> None:
        # Download, Preprocess files etc.
        pass

    def setup(self, tokenizer, batch_size, ...) -> None:
        with open(self.data_file_path, "r", encoding="utf-8") as file:
            data = json.load(file)

        # Partition the dataset into train and test
        train_data, test_data = random_split(data, ...)
        
        self.train_dataset = SFTDataset(data=train_data, ...)
        self.test_dataset = SFTDataset(data=test_data, ...)

    def train_dataloader(self) -> DataLoader:
        return DataLoader(self.train_dataset, ...)

    def val_dataloader(self) -> DataLoader:
        return DataLoader(self.test_dataset, ...)

    def test_dataloader(self) -> DataLoader:
        return self.val_dataloader()

```

It bundles:
- Preprocessing in `prepare_data()`
- Instantiation of datasets in `setup()`
- Definintion of dataloaders in `train/val/test_dataloader()`

**Note:** The `setup()` method takes special arguments as input that can't be deterimined immediately at the time of instantiating the datamodule. For exmaple, the tokenizer must be loaded from the checkpoint directory of the model, or the batch size is the micro-batch size set in the script.

## Usage in Training Scripts

The training scripts would simply add these lines of code (replacing the existing `get_batch()` function):

```py
datamodule = ...   # Instantiate from data CLI args
if fabric.global_rank == 0:
    datamodule.prepare_data()
fabric.barrier()
datamodule.setup(tokenizer, batch_size, ...)
train_dataloader = datamodule.train_dataloader()
val_dataloader = datamodule.val_dataloader()
train_dataloader, val_dataloader = fabric.setup_dataloaders(train_dataloader, val_dataloader)
```

A POC for this design is in #950.

## Pretraining
The pretraining datasets will still require preprocessing to be done externally due to their size. The corresponding datamodule would simply read the dataset at the default location, and error with instructions to preprocess if it can't be found.

## Extension

With this proposal, data preprocessing will be on-the-fly now. But we don't have that much to preprocess, since tokenization will be done as part of the dataloaders. The only real preprocessing we have is automatic creation of train-test splits for datasets that don't have them. This would be done in-memory with the proposal above, but as an extension, we could have a cache folder where we store the splits and only rerun it if the arguments change (hashed). This could be done as a follow up in the future.

## Pros

- Solves the aforementioned limitations
- It will be easier to write unit tests since everything will be more decoupled and modular

## Cons

- While this refactor can address the aforementioned limitations, it will introduce some level of abstraction (dataset + datamodule), which works slightly against the initial design philosophy of lit-gpt. The code is no longer in a single file that you can read top to bottom. This refactor will split things up into functions, dataset classes, etc. But we will try our best to keep the engineering effort to a minimum.

@carmocca @lantiga @rasbt @Andrei-Aksionov

### LINKED ISSUE #969: Data refactor follow-up work
After completing the main work in https://github.com/Lightning-AI/lit-gpt/issues/954, I need to look into:

- [x] Docstrings for all dataset parameters so they show up in the CLI help page.
- [x] Consistently use `max_seq_length` in `pretrain/tinyllama.py` as suggested by @carmocca 
- [x] Replace CSV data module with JSON datamodule: CSV is literally useless, like, as if your text data can't have any commas in it yeah sure (CSV uses a comma delimeter for columns)
- [x] Make prompt template configurable (both in the datasets and in all the scripts). Right now, they are hardcoded in the generate scripts and you are in serious trouble if you finetuned your model with template other than Alpaca and you don't change the hardcoded code.
- [x] Consider using data classes for data modules?
- [x] OpenWebText module for pretraining: This is a good reasonable small dataset for debugging pretraining. We should keep it but I need to reimplement to be able to use in the generic pretrain script (currently pretrain/tinyllama.py)
- [x] IOArgs is kinda redundant now, it only has 2 parameters in it that are not even supported by all scripts. We should probably just expose these at the top level in scripts where needed.

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #969
Fixes #954
Fixes #825

After this PR, you will be able to finetune

```
python finetune/full.py --config ...
```

and directly use the checkpoint:

```
python chat/base.py --checkpoint_dir out/finetune/full/final
```

Now you actually get the correct prompt style that you used for finetuning (e.g. Alpaca). On main, it does not apply the prompt style, you had to hard code it into chat.py, generate.py etc. if you wanted to use your checkpoint in Lit-GPT.

### CODE REVIEW

File: finetune/adapter.py [hunk 1]
```diff
@@ -229,6 +231,7 @@ def fit(
             if fabric.global_rank == 0:
                 copy_config_files(checkpoint_dir, checkpoint_file.parent)
                 save_hyperparameters(setup, checkpoint_file.parent)
+                save_prompt_style(data.prompt_style, checkpoint_file.parent)
```
Comment (awaelchli) [inline finetune/adapter.py:234]: The block of code above could probably move to a function. Or even a checkpointer. More like a checkpoint callback in, like in ...

<details>
<summary></summary>

<details>
<summary></summary>

<details>
<summary></summary>

<details>
<summary></summary>

<details>
<summary></summary>

**T     R     A     I     N     E     R**

</details>

</details>

</details>

</details>

</details>

**Review by carmocca — APPROVED** [review]
